Improve key derivation in RingCryptor

apyrgio opened the pull request:
This PR introduces the following changes to the RingCryptor interface:

1. It adds two public methods that allow the users to derive a key from a secret value and some metadata.
2. It removes key derivation from some low-level `RingCryptor` functions, while retaining it where it makes sense.

This should make key derivation less "auto-magic", and provide a solution to #6.

apyrgio committed:
cryptors: Add a helper that verifies key sizes

Add a helper associated function in `RingCryptor` that ensures that the
size of a key matches the encryption algorithm to be used. Currently,
this helper is used internally as a sanity check.

apyrgio committed:
cryptors: Factor out a common piece of code

Factor out a common piece of code that was copy-pasted in two places.

apyrgio committed:
cryptors: Add a helper that returns key sizes

Convert the private `RingCryptor._get_key_size()` method to a public
`RingCryptor::get_key_size()` associated function. Also, document it and
add an example in the docs.

apyrgio committed:
cryptors: Allow users to derive keys from metadata

While users had a way to derive a key from a secret value and some
PBKDF2 parameters, they didn't have a way to do so from a Metadata
struct.

Convert the `RingCryptor._derive_key()` method into a public associated
function called `derive_key_no_alloc()`, with minor changes. Also, add
some tests and documentation.

apyrgio committed:
cryptors: Add an easy and safe way to derive a key

Add a convenience wrapper around `derive_key_no_alloc` called
`derive_key`, which performs the key buffer allocation itself. Normally
this is dangerous [1], since it leaves the key in the heap and it may
remain there for a while. However, we dampen this risk by using the
`Zeroizing` trait from the `zeroize` [2] crate, which will zero the key
buffer before it's dropped.

[1]: https://crypto.stackexchange.com/questions/9998/should-i-delete-cryptographic-data-from-memory
[2]: https://crates.io/crates/zeroize

apyrgio committed:
cryptors: Derive keys only in specific methods

`RingCryptor` has the following public methods for
encryption/decryption:

* seal_with_passphrase
* seal_with_key
* seal_with_meta
* seal_in_place
* open
* open_with_meta
* open_in_place

In all of these methods, we previously used to perform key derivation.
The main reason was that the user could not do it themselves, but this
has been changed with the introduction of `derive_key`. On the other
hand, the problem with always performing key derivation internally is
that it can be slow, which is especially irritating when the same
passphrase is used successively. In these cases, it's better to allow
the user choose when and how they will derive the keys.

Looking into these methods, it seems that are two where key derivation
must remain:

1. seal_with_passphrase: In this method, we generate the key derivation
   metadata internally, so it's impossible to provide a derived key
   beforehand.
2. open: Callers of this method typically don't have access to the
   metadata struct, and thus can't know the key derivation metadata.

In all of the other methods, there is either no key derivation going on,
or they are more low-level. So, it makes sense to disable key derivation
in these cases, and expect that the user will pass the proper key.

Closes #6

apyrgio committed:
cryptors: Fix some typos

apyrgio committed:
cryptors: Remove a stale lint exception

apyrgio committed:
cryptors: Improve documentation on open_with_meta

apyrgio commented:
@djmitche, you can find the changes that we discussed on #6 on this PR. You can give it a try with:

```toml
tindercrypt = { git = "https://github.com/apyrgio/tindercrypt", branch = "feature-improve-key-deriv" }
```

The features you're interested in are:

1. You can use `let key = cryptor.derive_key(&meta, &pass)?;` to derive a key explicitly.
2. You can pass that key to `seal_with_meta` / `open_with_meta`, and no key derivation will take place.

If you don't have time to test this, that's perfectly fine, do it at your convenience. I'll leave this PR open until the weekend, and then merge it, but we can continue the discussion on #6 if necessary.

djmitche commented:
Thanks for flagging me!  I'll have a look now.  Sorry for the delay -- it's been a week.

djmitche commented:
This works for me!  The one complaint I might make is, it's odd to have to create a `RingCryptor` just to derive the key.  Skimming the `derive_key` function and the functions it calls, it seems none of them use properties of `self`.  Perhaps that could be changed to `RingCryptor::derive_key` or moved elsewhere in the crate?

I could just create a singleton RingCryptor, use it to derive the key once, and then use it to seal and open -- except that I'm using AAD for each transaction, which requires creating a new RingCryptor for each transaction.

As a workaround, it's easy enough to create a RingCryptor just to derive the key.

apyrgio commented:
> Thanks for flagging me! I'll have a look now. Sorry for the delay -- it's been a week.

No problem, and thanks a lot for the feedback.

> The one complaint I might make is, it's odd to have to create a RingCryptor just to derive the key. Skimming the derive_key function and the functions it calls, it seems none of them use properties of self. Perhaps that could be changed to RingCryptor::derive_key or moved elsewhere in the crate?

Right, and the same thing applies to `get_key_size()` / `verify_key_size()`. My intention was to make those functions `ring`-specific, since they use the `ring` crate underneath. So, I will convert them to associated functions, as you propose, given that `&self` is not used.

apyrgio commented:
Ok, so I've updated the branch with the changes you mentioned above. Basically, I've converted all newly introduced methods to associated functions, since `&self` was not used. If the changes look good to you, I can merge it.

djmitche reviewed (commented):
Looks good!

apyrgio commented:
Great, thanks again for the feedback. I'll create a 0.3 release soon.
